docs/proposals/config_data.md

```diff
@@ -0,0 +1,262 @@
+```json
+{
+    "apiVersion": "v1",
+    "name": "PodConfiguration",
+    "kind": "ConfigData",
+    "data": {
+        "API_SERVER": "10.240.13.14:8998",
+        "API_VERSION": "v1",
+        "DB_HOST": "10.240.55.3",
+        "DB_PORT": "3306"
+    }
+}
+```
+
+To persist this ConfigData object to the API server, use the following command:
+```bash
+kubectl create -f pod-config.json
+```
+
+Now, the ``PodConfiguration`` ConfigData object is stored in ``etcd``.
+The new object can be accessed, referenced, used and updated through the API server.
+
+## Configuration Workflow
+
+The component configuration workflow is given below:
+
+1. ``ConfigData`` object is created through the API server.
+2. New components using the ``ConfigData`` object are created.
+3. ``ConfigData`` object is retrieved using the downward API.
+4. ``Data`` key/value pairs of ``ConfigData`` will be exposed as either environment variables (``EnvVarSource``) or as a new volume (``ConfigDataVolumeSource``).
+5. API server is watched for modifications to the object.
+6. After a modification, actions 3, 4, and 5 are repeated.
+
+**Note**: Environment variables specified in ``ConfigData`` will override the environment variables
+with the same name found in the system if consumed as environment variables.
+
+## Use Cases
+
+#### Pod Configuration
+
+```json
+{
+    "apiVersion": "v1",
+    "name": "PodConfiguration1",
+    "kind": "ConfigData",
+    "data": {
+        "API_SERVER": "10.240.13.14:8998",
+        "API_VERSION": "v1",
+        "DB_HOST": "10.240.55.3",
+        "DB_PORT": "3306"
+    }
+}
+```
+``PodConfiguration1`` is intended to be consumed as environment variable key/value pairs.
+
+```json
+{
+    "apiVersion": "v1",
+    "name": "PodConfiguration2",
+    "kind": "ConfigData",
+    "data": {
+        "config.yaml": "CONFIG_YAML_STRING_CONTENTS",
+        "spec.yaml": "SPEC_YAML_STRING_CONTENTS"
+    }
+}
+```
+``PodConfiguration2`` is intended to be used as a mounted volume containing two files with their respective contents.
+
+Then, configuration objects are persisted to the API server as follows:
+```bash
+kubectl create -f pod-configuration-1.json 
+kubectl create -f pod-configuration-2.json
+```
+
+We create a replication controller specification that creates pods using these two configuration objects:
+```yaml
+apiVersion: v1
+kind: ReplicationController
+metadata:
+  name: ExampleReplicationController
+spec:
+  replicas: 2
+  selector:
+    app: dummy-app
+  template:
+    metadata:
+      labels:
+        app: dummy-app
+    spec:
+      # New pods will be created using the ConfigData objects.
+      # Each config object will be updated be it a volume or environment variables.
+      # But, some of them might require a restart on update.
+      # Note: If multiple env ConfigData objects are specified, let's call them Env1 and Env2.
+      #       If both of them contains a same named environment variable, Env2 will be the final one.
```

Comment:
You would want to be able to map the environment variable.  Just because
config has PASSWORD=secret, I may want that to be "DB_PASSWORD=secret" when
I use it in my pod.  Downward pod API is about adapting the Kube world into
things that an image that doesn't know about Kube might need.

On Mon, Sep 14, 2015 at 5:16 PM, Tamer TAS <notifications@github.com> wrote:

> In docs/proposals/config_data.md
> <https://github.com/kubernetes/kubernetes/pull/6477#discussion_r39448300>:
>
> > +metadata:
> > +  name: ExampleReplicationController
> > +spec:
> > +  replicas: 2
> > +  selector:
> > +    app: dummy-app
> > +  template:
> > +    metadata:
> > +      labels:
> > +        app: dummy-app
> > +    spec:
> > +      # New pods will be created using the ConfigData objects.
> > +      # Each config object will be updated be it a volume or environment variables.
> > +      # But, some of them might require a restart on update.
> > +      # Note: If multiple env ConfigData objects are specified, let's call them Env1 and Env2.
> > +      #       If both of them contains a same named environment variable, Env2 will be the final one.
>
> The ConfigData object will contain multiple fields (in this case env
> vars). Wouldn't describing each env var as it is used normally in the
> downward API for single ones unneccessary?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/6477/files#r39448300>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift